File: neet_code_all/src/array_hash/can_place_flowers.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn can_place_flowers(mut flowerbed: Vec<i32>, mut n: i32) -> bool {
        let len = flowerbed.len();

        if len == 1 {
            if flowerbed[0] == 0 {
                return n - 1 == 0;
            }
            return n == 0;
        }

        for i in 0..len {
            if i == 0 {
                if flowerbed[i] == 0 && flowerbed[i + 1] == 0 {
                    flowerbed[i] = 1;
                    n -= 1;
                }
            } else if i == len - 1 {
                if flowerbed[i] == 0 && flowerbed[i - 1] == 0 {
                    flowerbed[i] = 1;
                    n -= 1;
                }
            } else {
                if flowerbed[i - 1] == 0 && flowerbed[i] == 0 && flowerbed[i + 1] == 0 {
                    flowerbed[i] = 1;
                    n -= 1;
                }
            }

            if n == 0 {
                println!("flowerbed: {:?}", flowerbed);
                return true;
            }
        }

        println!("flowerbed: {:?}", flowerbed);
        n == 0
    }

    pub fn can_place_flowers_2(mut flowerbed: Vec<i32>, mut n: i32) -> bool {
        let len = flowerbed.len();

        if len == 1 {
            if flowerbed[0] == 0 {
                return n - 1 == 0;
            }
            return n == 0;
        }

        let mut l = 0;
        while l < len && n != 0 {
            if l == 0 {
                if flowerbed[l] == 0 && flowerbed[l + 1] == 0 {
                    flowerbed[l] = 1;
                    n -= 1;
                }
            } else if l == len - 1 {
                if flowerbed[l] == 0 && flowerbed[l - 1] == 0 {
                    flowerbed[l] = 1;
                    n -= 1;
                }
            } else {
                if flowerbed[l - 1] == 0 && flowerbed[l] == 0 && flowerbed[l + 1] == 0 {
                    flowerbed[l] = 1;
                    n -= 1;
                }
            }

            l += 1;
        }

        n == 0
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    // time O(N), space: O(1)
    pub fn can_place_flowers_1(flowerbed: Vec<i32>, n: i32) -> bool {
        let mut empty = if flowerbed[0] != 0 { 0 } else { 1 };
        let mut n = n;

        for f in flowerbed {
            if f != 0 {
                n -= (empty - 1) / 2;
                empty = 0;
            } else {
                empty += 1;
            }
        }

        n -= empty / 2;

        n <= 0
    }

    // time O(N), space: O(1)
    pub fn can_place_flowers_2(mut flowerbed: Vec<i32>, mut n: i32) -> bool {
        for i in 0..flowerbed.len() {
            if n == 0 {
                return true;
            }

            if (i == 0 || flowerbed[i - 1] == 0)
                && flowerbed[i] == 0
                && (i == flowerbed.len() - 1 || flowerbed[i + 1] == 0)
            {
                flowerbed[i] = 1;
                n -= 1;
            }
        }

        n == 0
    }

    // time O(N), space: O(N)
    pub fn can_place_flowers_3(flowerbed: Vec<i32>, mut n: i32) -> bool {
        let mut f: Vec<i32> = vec![vec![0], flowerbed, vec![0]]
            .into_iter()
            .flatten()
            .collect();

        for i in 1..f.len() - 1 {
            if f[i - 1] == 0 && f[i] == 0 && f[i + 1] == 0 {
                f[i] = 1;
                n -= 1;
            }
        }

        n <= 0
    }
}

fn main() {
    let case_1 = (vec![1, 0, 0, 0, 1], 1);
    // => true
    let case_2 = (vec![1, 0, 0, 0, 1], 2);
    // => false
    let case_3 = (vec![1, 0, 0, 0, 0, 1], 2);
    // => false
    let case_4 = (vec![0, 0, 1, 0, 1], 1);
    // => true
    let case_5 = (vec![1, 0, 0, 0, 1, 0, 0], 2);
    // => true
    let case_6 = (vec![0, 0, 1, 0, 0], 1);
    // => true

    println!(
        "case_3: {:?}",
        SolutionAns::can_place_flowers_1(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_3: {:?}",
        SolutionAns::can_place_flowers_2(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_3: {:?}",
        SolutionAns::can_place_flowers_3(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_4: {:?}",
        SolutionAns::can_place_flowers_1(case_4.0.clone(), case_4.1)
    );
    println!(
        "case_4: {:?}",
        SolutionAns::can_place_flowers_2(case_4.0.clone(), case_4.1)
    );
    println!(
        "case_4: {:?}",
        SolutionAns::can_place_flowers_3(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_5: {:?}",
        SolutionAns::can_place_flowers_1(case_5.0.clone(), case_5.1)
    );
    println!(
        "case_5: {:?}",
        SolutionAns::can_place_flowers_2(case_5.0.clone(), case_5.1)
    );
    println!(
        "case_5: {:?}",
        SolutionAns::can_place_flowers_3(case_5.0.clone(), case_5.1)
    );
}

File: neet_code_all_cpp/array_hash/can_place_flowers.cpp
#include <cinttypes>
#include <iostream>
#include <vector>

// 模範解答
class SolutionAns {
    public:
        bool canPlaceFlowers(std::vector<int>& flowerbed, int n) {
            if(n == 0) {
                return true;
            }

            // 頭とケツに0を余計に追加
            flowerbed.insert(flowerbed.begin(), 0);
            flowerbed.push_back(0);

            for(int i = 1; i < flowerbed.size() - 1; ++i) {
                // 隣り合った3個がゼロなら1に変更
                if(flowerbed[i - 1] == 0 && flowerbed[i] == 0 && flowerbed[i + 1] == 0) {
                    flowerbed[i] = 1;
                    --n;
                }

                if(n == 0) {
                    return true;
                }
            }

            return false;
        }

};

// Rustの模範解答より
class SolutionAnsRust {
    public:
        // time O(N), space: O(1)
        bool canPlaceFlowers(std::vector<int>& flowerbed, int n) {
            int empty = 0;
            if(!flowerbed[0]) {
                empty = 1;
            }

            for(auto& f: flowerbed) {
                if(f) {
                    n -= (empty - 1) / 2;
                    empty = 0;
                } else {
                    empty += 1;
                }
            }

            n -=  empty / 2;

            return n <= 0;
        }

        // time O(N), space: O(1)
        bool canPlaceFlowers2(std::vector<int> flowerbed, int n) {
            for(int i = 0; i < flowerbed.size(); ++i) {
                if(n == 0) {
                    return true;
                }

                if((i == 0 || flowerbed[i - 1] == 0)
                        && flowerbed[i] == 0
                        && (i == flowerbed.size() - 1 || flowerbed[i + 1] == 0)
                        ) 
                {
                    flowerbed[i] = 1;
                    --n;
                }
            }

            return n == 0;
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, int>;

    Case case_1 = {{1, 0, 0, 0, 1}, 1};
    // => true
    Case case_2 = {{1, 0, 0, 0, 1}, 2};
    // => false
    Case case_3 = {{0, 0, 1, 0, 1}, 1};
    // => true
    Case case_4 = {{0, 0, 1, 0, 0}, 1};
    // => true

    SolutionAnsRust s_ans_rs;

    std::cout << "case_1: " << s_ans_rs.canPlaceFlowers(case_1.first, case_1.second) << std::endl;
    std::cout << "case_2: " << s_ans_rs.canPlaceFlowers(case_2.first, case_2.second) << std::endl;
    std::cout << "case_3: " << s_ans_rs.canPlaceFlowers(case_3.first, case_3.second) << std::endl;
    std::cout << "case_4: " << s_ans_rs.canPlaceFlowers(case_4.first, case_4.second) << std::endl;

    std::cout << "case_1: " << s_ans_rs.canPlaceFlowers2(case_1.first, case_1.second) << std::endl;
    std::cout << "case_2: " << s_ans_rs.canPlaceFlowers2(case_2.first, case_2.second) << std::endl;
    std::cout << "case_3: " << s_ans_rs.canPlaceFlowers2(case_3.first, case_3.second) << std::endl;
    std::cout << "case_4: " << s_ans_rs.canPlaceFlowers2(case_4.first, case_4.second) << std::endl;

    SolutionAns s_ans;

    std::cout << "case_1: " << s_ans.canPlaceFlowers(case_1.first, case_1.second) << std::endl;
    std::cout << "case_2: " << s_ans.canPlaceFlowers(case_2.first, case_2.second) << std::endl;
    std::cout << "case_3: " << s_ans.canPlaceFlowers(case_3.first, case_3.second) << std::endl;
    std::cout << "case_4: " << s_ans.canPlaceFlowers(case_4.first, case_4.second) << std::endl;
}
